Added FanOut/FanIn Pattern (#1800)
iluwatar/java-design-patterns: fanout-fanin/README.md

---
layout: pattern
title: Fan-Out/Fan-In
folder: fanout-fanin
permalink: /patterns/fanout-fanin/
categories: Integration
language: en
tags:
- Microservices
---

## Intent

> The pattern is used when a source system needs to run one or more long-running process that will fetch some data. 

Reviewer: ```suggestion
The pattern is used when a source system needs to run one or more long-running processes that will fetch some data. 
```

> Source will not block itself waiting for the reply. <br> The pattern will run the same function in multiple 

Reviewer: ```suggestion
The source will not block itself waiting for the reply. <br> The pattern will run the same function in multiple 
```

services or machines to fetch the data. This is equivalent to invoking the function multiple times on different chunks of data.  

## Explanation

> The FanOut/FanIn service will take in a list of requests and a consumer. Each request might complete at different time.

Reviewer: ```suggestion
The FanOut/FanIn service will take in a list of requests and a consumer. Each request might complete at a different time.
```

FanOut/FanIn service will accept the input params and returns the initial system an ID to acknowledge that the pattern
service has received the requests. Now the caller will not wait or expect the result in the same connection. 

Meanwhile, the pattern service will invoke the requests that have come. The requests might complete at different time. 
These requests will be processed in different instances of the same function in different machines or services. As the 

> requests get completed, a callback service everytime is called that transforms the result into a common format, or a single object 

Reviewer: ```suggestion
requests get completed, a callback service is called that transforms the result into a common single object format
```

> that gets pushed to a consumer. That caller will be at the other end of the consumer receiving the result.

Reviewer: ```suggestion
that gets pushed to a consumer. The caller will be at the other end of the consumer receiving the result.
```

> 

Reviewer: Please add a programmatic example in the explanation section. See the other patterns for examples of how to do it.

## Class diagram
![alt-text](./etc/fanout-fanin.png)

## Applicability

> Use this pattern when you can chunk the workload or load into multiple chunks that can be dealt with separately.

Reviewer: ```suggestion
Use this pattern when you can divide the workload into multiple chunks that can be dealt with separately.
```

## Related patterns

- Aggregator Microservices

> - API Gateway

Reviewer: Add links
